qa/tasks/ceph2.py

```diff
@@ -0,0 +1,824 @@
+            teuthology.write_file(
+                remote=remote,
+                path='{}/{}.keyring'.format(testdir, cluster_name),
+                data=ctx.ceph[cluster_name].admin_keyring)
+
+            log.info('Adding host %s to orchestrator...' % remote.shortname)
+            shell(ctx, cluster_name, remote, [
+                'ceph', 'orchestrator', 'host', 'add',
+                remote.shortname
+            ])
+
+        yield
+
+    finally:
+        log.info('Cleaning up testdir ceph.* files...')
+        ctx.cluster.run(args=[
+            'rm', '-f',
+            '{}/seed.{}.conf'.format(testdir, cluster_name),
+            '{}/{}.pub'.format(testdir, cluster_name),
+            '{}/{}.conf'.format(testdir, cluster_name),
+            '{}/{}.keyring'.format(testdir, cluster_name),
+        ])
+
+        log.info('Stopping all daemons...')
+
+        # this doesn't block until they are all stopped...
+        #ctx.cluster.run(args=['sudo', 'systemctl', 'stop', 'ceph.target'])
+
+        # so, stop them individually
+        for role in ctx.daemons.resolve_role_list(None, CEPH_ROLE_TYPES):
+            cluster, type_, id_ = teuthology.split_role(role)
+            ctx.daemons.get_daemon(type_, id_, cluster).stop()
+
+@contextlib.contextmanager
+def ceph_mons(ctx, config):
+    """
+    Deploy any additional mons
+    """
+    cluster_name = config['cluster']
+    fsid = ctx.ceph[cluster_name].fsid
+    testdir = teuthology.get_testdir(ctx)
+    num_mons = 1
+
+    try:
+        for remote, roles in ctx.cluster.remotes.items():
+            for mon in [r for r in roles
+                        if teuthology.is_type('mon', cluster_name)(r)]:
+                c_, _, id_ = teuthology.split_role(mon)
+                if c_ == cluster_name and id_ == ctx.ceph[cluster_name].first_mon:
+                    continue
+                log.info('Adding %s on %s' % (mon, remote.shortname))
+                num_mons += 1
+                shell(ctx, cluster_name, remote, [
+                    'ceph', 'orchestrator', 'mon', 'update',
+                    str(num_mons),
+                    remote.shortname + ':' + ctx.ceph[cluster_name].mons[mon] + '=' + id_,
+                ])
+                ctx.daemons.register_daemon(
+                    remote, 'mon', id_,
+                    cluster=cluster_name,
+                    fsid=fsid,
+                    logger=log.getChild(mon),
+                    wait=False,
+                    started=True,
+                )
+
+                while True:
+                    log.info('Waiting for %d mons in monmap...' % (num_mons))
+                    r = shell(
+                        ctx=ctx,
+                        cluster_name=cluster_name,
+                        remote=remote,
+                        args=[
+                            'ceph', 'mon', 'dump', '-f', 'json',
+                        ],
+                        stdout=StringIO(),
```

Comment:
Can the 'shell' can be refactored to use 'remote.sh' instead of 'remote.run' otherwise the code should be refactored to not use cStringIO and be py3 compatible.